src/pyhf/infer/mixins.py

```diff
@@ -0,0 +1,116 @@
+"""
+Mixins for Hypothesis Testing.
+"""
+from .. import get_backend
+from .test_statistics import qmu
+from .calculators import AsymptoticTestStatDistribution, EmpiricalDistribution
+
+
+class Calculator(object):
+    def __init__(
+        self, data, pdf, init_pars=None, par_bounds=None, qtilde=False, ntoys=2000
+    ):
+        self.data = data
+        self.pdf = pdf
+        self.init_pars = init_pars or pdf.config.suggested_init()
+        self.par_bounds = par_bounds or pdf.config.suggested_bounds()
+        self.qtilde = qtilde
+        self.distribution = None
+
+        # TODO: better names???
+        # for Asymptotics, it is self.sqrtqmuA_v
+        # for Toys, it is signal/bkg qtilde
+        self.something_signal = None
+        self.something_bkg = None
+
+        # toys
+        self.ntoys = ntoys
+
+    def distributions(self, poi_test):
+        if self.something_signal is None or self.something_bkg is None:
+            raise RuntimeError('need to call .teststatistic(poi_test) first')
+
+        if self.distribution is None:
+            raise RuntimeError('need to call this from a mixin\'d class')
+
+        s_plus_b = self.distribution(signal_qtilde)
```

Comment:
I don't think we have a guarantee that the init signature of distribution have the same semantics so I wouldn't add this here